standardize bot comments

crimsonfaith91 opened the pull request:
addressing https://github.com/kubernetes/test-infra/issues/2820

All test cases were passed for change of `@k8s-bot test this` command to `/test`.

ixdy commented on prow/plugins/yuks/yuks.go:
```diff
@@ -29,7 +29,7 @@ import (
 
 var (
-	match  = regexp.MustCompile(`(?m)^@k8s-(ci|ro)?bot tell me a joke[,.?!]*\r?$`)
+	match  = regexp.MustCompile(`/tell-me-a-joke`) //regexp.MustCompile(`(?m)^@k8s-(ci|ro)?bot tell me a joke[,.?!]*\r?$`)
```
bikeshed: I'd probably go with just `/joke` instead.

crimsonfaith91 replied:
i will change it to `/joke` :)

ixdy commented:
@k8s-bot ok to test

crimsonfaith91 commented:
@ixdy noted! thanks for the `ok to test`!

crimsonfaith91 commented:
Based on `config.yaml` file:

For `pull-security-kubernetes-verify`, the old trigger command is `@k8s-bot (pull-security-kubernetes-verify |verify )?test this`.
For `pull-security-kubernetes-e2e-gce`, the old trigger command is `@k8s-bot pull-security-kubernetes-e2e-gce test this`.

Is regex necessary? Which one should i follow? Thanks!

crimsonfaith91 commented:
All test cases were passed for change of `@k8s-bot test this` command to `/test`.
i left the different trigger command intact.

crimsonfaith91 commented:
@k8s-ci-robot test this

crimsonfaith91 commented:
@k8s-bot tell me a joke

crimsonfaith91 commented:
@k8s-bot test this

crimsonfaith91 commented:
The new bot comments will be as follow:
(1) `/test` = `@k8s-bot test this`
(2) `/test pull-kubernetes-verify` = `@k8s-bot pull-kubernetes-verify test this`
(3) `/ok-to-test` = `@k8s-bot ok to test`
(4) `/joke` = `@k8s-bot tell me a joke`

I have to skip `CheckRetest` in `prow/config/config_test.go` to ensure its GitHub version passed. Due to bot command change, the test will not passed on GitHub as it is a retest. However, it is still passed locally.

crimsonfaith91 commented on prow/plugins/yuks/yuks.go:
```diff
@@ -29,7 +29,7 @@ import (
 
 var (
-	match  = regexp.MustCompile(`(?m)^@k8s-(ci|ro)?bot tell me a joke[,.?!]*\r?$`)
+	match  = regexp.MustCompile(`/joke`) //regexp.MustCompile(`(?m)^@k8s-(ci|ro)?bot tell me a joke[,.?!]*\r?$`)
```
i will remove the extra comment at end of line.

crimsonfaith91 commented:
/unassign @grodrigues3

crimsonfaith91 commented:
/assign @spxtr 
Do you mind to review this PR? Thanks!

cblecker commented:
@crimsonfaith91 needs a rebase

additionally, this change should probably have an e-mail out to kubernetes-dev@ when it gets merged as it will change basic workflows around testing

fejta commented:
This is fantastic!

However we should preserve the existing test commands. In particular we cannot regress @k8s-bot test this which is something the submit queue uses to request testing. 

Also we should first update docs to recommend the /test foo command while still supporting @k8s-bot foo test this. A few weeks after that change we can consider dropping @k8s-bot foo test this support. 

This will also give us time to release a new submit queue that uses /test

/assign

crimsonfaith91 commented:
@cblecker yup! i will send an email to kubernetes-dev@ when this PR gets merged. :)

crimsonfaith91 commented:
@fejta Definitely! i will change my codes to accommodate both new and old bot comments for a few weeks, and then create a new PR removing the old ones. Thank you very much for your detailed explanation! :)

crimsonfaith91 commented:
/unassign spxtr

crimsonfaith91 commented:
/unassign @spxtr

crimsonfaith91 commented:
The codes have been rebased. Both `@k8s-bot test this` and `/test` work for trigger commands. However, rerun commands will still use the old syntax `@k8s-bot test this` to ensure retests are passing. Any feedback regarding this?

fejta commented on prow/config.yaml:
```diff
@@ -19,35 +19,35 @@ presubmits:
     context: pull-cadvisor-e2e
     rerun_command: "@k8s-bot pull-cadvisor-e2e test this"
-    trigger: "@k8s-bot (pull-cadvisor-e2e )?test this"
+    trigger: "(^@k8s-bot (pull-cadvisor-e2e )?test this$|^/test( pull-cadvisor-e2e)?$)"
```
Most REs seem to have a `\r?$` pattern, not sure why (ask @spxtr) maybe something about windows?

https://github.com/kubernetes/test-infra/blob/master/prow/plugins/assign/assign.go#L33

crimsonfaith91 replied:
sure!

crimsonfaith91 replied:
it is for handling `\r\n`. IMO, we can have `\r?$` for `trigger`, but not `rerun_command`.

fejta commented on prow/config.yaml:
```diff
@@ -19,35 +19,35 @@ presubmits:
     always_run: true
     context: pull-cadvisor-e2e
     rerun_command: "@k8s-bot pull-cadvisor-e2e test this"
```
You should change the suggested `rerun_command` to `/test pull-cadvisor-e2e`.

Also, please write unit tests that validate the following continue to work:
Already exist:
`@k8s-bot JOB test this`
`@k8s-got test this`  
Needed
`/test JOB`
`/test`

crimsonfaith91 replied:
noted! i will update `rerun_command`, and add more test cases for `/test`.

fejta commented on prow/plugins/trigger/ic.go:
```diff
@@ -26,7 +26,7 @@ import (
 )
 
-var okToTest = regexp.MustCompile(`(?m)^(@k8s-bot )?ok to test\r?$`)
+var okToTest = regexp.MustCompile(`/ok-to-test`)
```
This should a) start recommending `/ok-to-test` but continue to accept `@k8s-bot ok to test` for awhile.

Also please add unit tests to validate that both `@k8s-bot ok to test` and `/ok-to-test` work

crimsonfaith91 replied:
noted! i thought we only need to have `/test` to work alongside with `@k8s-bot test this` for a few weeks. it makes sense to have all bot comments doing so. i will also add test cases for `/joke`.

fejta commented:
rerun_command is the example command the bot uses to tell the user how to rerun the test. This should be updated.

What we need is to retire the old pattern from the regexp a few weeks after the rerun command is updated.

Also note that the submit queue requests testing using the @k8s-bot test this pattern, which will need to get updated.

crimsonfaith91 commented:
@fejta Thanks for catching the submit queue part! Along with this change, i will update the retest codes found at `config_test.go` accordingly too.

fejta commented:
Cool let me know when this is ready for a review :grin:

crimsonfaith91 commented:
@fejta The codes now handle both new and old syntax with test cases for both. The change on submit queue is not trivial. I am still working at it.

crimsonfaith91 commented on prow/config.yaml:
```diff
@@ -592,28 +592,28 @@ presubmits:
-    rerun_command: "@k8s-bot pull-security-kubernetes-node-e2e test this"
-    trigger: "@k8s-bot (pull-security-kubernetes-node-e2e |node e2e )?test this"
+    rerun_command: "/test pull-security-kubernetes-node-e2e"
+    trigger: "(^@k8s-bot (pull-security-kubernetes-node-e2e |node e2e )?test this\r?$|^/test( pull-security-kubernetes-node-e2e| node e2e)?\r?$)"
```
Yup! It will simplify the regex a lot. :)
Should i remove _all_ alternatives?

fejta replied:
Sure so just support the following:
```
// Matches job name
@k8s-bot pull-security-kubernetes-node-e2e test this
/test pull-security-kubernetes-node-e2e
// If always_run: true then also support:
/test
@k8s-bot test this
```

crimsonfaith91 replied:
Noted! If `always_run` equals to false, it will not support `/test` and `k8s-bot test this`.

crimsonfaith91 commented on prow/config.yaml:
```diff
@@ -592,28 +592,28 @@ presubmits:
-    rerun_command: "@k8s-bot pull-test-infra-bazel test this"
-    trigger: "@k8s-bot (pull-test-infra-bazel |bazel )?test this"
+    rerun_command: "/test pull-test-infra-bazel"
+    trigger: "(^@k8s-bot (pull-test-infra-bazel |bazel )?test this\r?$|^/test( pull-test-infra-bazel| bazel)?\r?$)"
```
sure! i will remove `/test bazel`, `/test pull-test-infra-bazel`, and `@k8s-bot pull-test-infra-bazel test this`.

crimsonfaith91 commented on prow/config.yaml:
```diff
@@ -678,8 +678,8 @@ presubmits:
-    rerun_command: "@k8s-bot pull-test-infra-verify-bazel test this"
-    trigger: "@k8s-bot (pull-test-infra-verify-bazel |verify )?test this"
+    rerun_command: "/test pull-test-infra-verify-bazel"
+    trigger: "(^@k8s-bot (pull-test-infra-verify-bazel |verify )?test this\r?$|^/test( pull-test-infra-verify-bazel| verify)?\r?$)"
```
ya! i will remove `/test verify`, `/test pull-test-infra-verify-bazel`, and `@k8s-bot pull-test-infra-verify-bazel test this`.

crimsonfaith91 commented:
@fejta Addressed feedback.
For submit queue that requests testing, how should it handle both new and old commands? There are 2 tests that still require `@k8s-bot test this`: `bazel` and `verify` tests.

fejta commented on prow/config.yaml:
```diff
@@ -592,28 +592,28 @@ presubmits:
     always_run: true
-    rerun_command: "@k8s-bot pull-test-infra-bazel test this"
-    trigger: "@k8s-bot (pull-test-infra-bazel |bazel )?test this"
+    rerun_command: "@k8s-bot bazel test this"
```
Why doesn't this one follow the same pattern as all the others?

fejta commented on prow/config.yaml:
```diff
@@ -678,8 +678,8 @@ presubmits:
     - master
-    rerun_command: "@k8s-bot pull-test-infra-verify-bazel test this"
-    trigger: "@k8s-bot (pull-test-infra-verify-bazel |verify )?test this"
+    rerun_command: "@k8s-bot verify test this"
```
Why doesn't this one follow the same pattern as all the others?

fejta commented on prow/config.yaml:
```diff
@@ -592,28 +592,28 @@ presubmits:
-    rerun_command: "@k8s-bot pull-test-infra-bazel test this"
-    trigger: "@k8s-bot (pull-test-infra-bazel |bazel )?test this"
+    rerun_command: "@k8s-bot bazel test this"
+    trigger: "^@k8s-bot bazel test this\r?$"
```
Something is very broken if this has `always_run: true`, does not trigger on `@k8s-bot test this` and yet unit tests pass!

fejta commented on prow/config/jobs_test.go:
```diff
@@ -76,8 +74,22 @@ func TestPresubmits(t *testing.T) {
+			if job.AlwaysRun {
+				var testCmd string
+				cmdMap := make(map[string]string)
+				cmdMap["pull-test-infra-bazel"] = "bazel"
```
Why do we have this? These two jobs should work the same way as everything else

fejta commented on prow/config/jobs_test.go:
```diff
@@ -76,8 +74,22 @@ func TestPresubmits(t *testing.T) {
-			if job.AlwaysRun && !job.re.MatchString(testThis) {
-				t.Errorf("AlwaysRun job %s: \"%s\" does not match regex \"%v\".", job.Name, testThis, job.Trigger)
+
+			if job.AlwaysRun {
```
If job.Always run is true then both `/test` and `@k8s-bot test this` need to trigger the job.

All jobs need to trigger on `/test JOB` and `@k8s-bot JOB test this`.

crimsonfaith91 commented:
@fejta From my understanding, for both the tests, we will still use the old `@k8s-bot test this` comment. For example, i will remove `/test bazel`, `/test pull-test-infra-bazel`, and `@k8s-bot pull-test-infra-bazel test this`. As a result, only `@k8s-bot bazel test this` remains for usage.

Did you mean `^@k8s-bot( bazel)? test this$`?

fejta commented:
No, there is no job named bazel, so `@k8s-bot bazel test this` should not trigger anything.

If a job is set to `always_run: true`, then we need to configure it to trigger on a `/test` and a `@k8s-bot test this comment` and we need unit tests to validate that this is the case.

We also need to validate that both `@k8s-bot JOB test this` and `/test JOB` trigger each job.

What we do not want are weird/zombie variants of trigger a job `@k8s-bot random string test this` or `/test random string`.

`/test` and `@k8s-bot test this` should match if **and only if** the job has `always_run: true`. Otherwise the only way to trigger the job should be `/test JOB_NAME` or `@k8s-bot JOB_NAME test this` and the example rerun_command should suggest the new `/test JOB_NAME` format, not the old `@k8s-bot JOB_NAME test this` format

crimsonfaith91 commented:
@fejta Thanks for the elaborated explanation towards how the trigger and rerun commands are working. i was confused when you mentioned keeping `@k8s-bot bazel test this` in one of the comments above. Did you mean `@k8s-bot pull-test-infra-bazel test this`?

To clarify, for `pull-test-infra-bazel` test, we should have code snippet below:
```
always_run: true
rerun_command: "/test pull-test-infra-bazel"
trigger: "(^@k8s-bot (pull-test-infra-bazel )?test this\r?$|^/test( pull-test-infra-bazel)?\r?$)"
```
Am i correct? Sorry for the inconvenience caused.

fejta commented:
Yes, exactly! And same thing for all the other jobs. Thanks for iterating with me on this

crimsonfaith91 commented:
@fejta updated the codes. the submit queue of mungegithub now uses `/test`.

fejta commented on prow/config/jobs_test.go:
```diff
@@ -95,7 +95,7 @@ func TestPresubmits(t *testing.T) {
 				}
 				if job2.re.MatchString(job.RerunCommand) {
-					t.Errorf("RerunCommand \"%s\" from job %s matches \"%v\" from job %s but shouldn't.", job.RerunCommand, job.Name, job2.Trigger, job2.Name)
+					t.Errorf("RerunCommand \"%s\" from job %s matches \"%v\" from job %s but shouldn't.", job.RerunCommand, job.Name, job2.Trigger, job2.Name) //TODO
```
What is the TODO?

crimsonfaith91 replied:
it was the comment to remind me which part of codes that i need to change. i forgot to remove it afterwards...

fejta commented on prow/config/config_test.go:
```diff
@@ -82,7 +82,7 @@ func TestContextMatches(t *testing.T) {
 func CheckRetest(t *testing.T, repo string, presubmits []Presubmit) {
 	for _, p := range presubmits {
-		expected := fmt.Sprintf("@k8s-bot %s test this", p.Name)
+		expected := fmt.Sprintf("/test %s", p.Name)
```
This is good as we validate that `/test JOB` works. However we need something here that validates that `@k8s-bot JOB test this` continues to work, which we no longer have.

crimsonfaith91 replied:
ya too!

crimsonfaith91 replied:
updated: we will only handle `/test %s`.

fejta commented on prow/config/jobs_test.go:
```diff
@@ -28,7 +28,7 @@ import (
 var podRe = regexp.MustCompile(`^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`)
 
-const testThis = "@k8s-bot test this"
+const testThis = "/test"
```
Likewise, this change makes it that we validate that `/test` will trigger all the `always_run: true` jobs (and only those jobs).

However we also need to still validate that `@k8s-bot test this` triggers all the `always_run: true` jobs (and none of the others).

We no longer have this coverage (feel free to test both of these facts)

crimsonfaith91 replied:
yup! i should still need to add the test cases for the old commands.

crimsonfaith91 commented on prow/plugins/yuks/yuks_test.go:
```diff
@@ -94,6 +99,7 @@ func TestJokes(t *testing.T) {
 		shouldComment bool
 		shouldError   bool
 	}{
+		/* Old syntax */
```
noted!

crimsonfaith91 commented on prow/plugins/yuks/yuks.go:
```diff
@@ -29,7 +29,7 @@ import (
 
 var (
-	match  = regexp.MustCompile(`(?m)^@k8s-(ci|ro)?bot tell me a joke[,.?!]*\r?$`)
+	match  = regexp.MustCompile(`((?m)^@k8s-(ci|ro)?bot tell me a joke[,.?!]*\r?$|^/joke\r?$)`)
```
yup, i will remove the old joke

crimsonfaith91 commented on prow/config/jobs_test.go:
```diff
@@ -162,42 +183,93 @@ func TestCommentBodyMatches(t *testing.T) {
 				{
 					Name:      "gce",
-					re:        regexp.MustCompile(`@k8s-bot (gce )?test this`),
+					re:        regexp.MustCompile(`(^@k8s-bot (gce )?test this\r?$|^/test( gce)?\r?$)`),
```
but if we remove the regex for old syntax, all the old syntax test cases above will not pass. in other words, the old syntax test cases will not be tested.

fejta replied:
That's why I'm suggesting you think through what you're actually testing.

**for this test function** what additional coverage are you adding via the old syntax test cases?

crimsonfaith91 replied:
i think i got what you mean. The function tests matches in comment body. It is not needed to test whether it matches all variants (new and old) of the regex as they are fake. Furthermore, it adds no test coverage to functions in `jobs.go`. correct me if i am wrong.

crimsonfaith91 commented:
@fejta i updated the codes based on previous feedback - it is now handling only commands like `/something gce` when testing comment body matches for presubmits. the commands include both random words and `/test`.

fejta commented on prow/plugins/trigger/ic_test.go:
```diff
@@ -174,14 +282,14 @@ func TestHandleIssueComment(t *testing.T) {
 					AlwaysRun: true,
 					Context:   "pull-job",
-					Trigger:   "@k8s-bot test this",
+					Trigger:   "(^@k8s-bot test this\r?$|^/test\r?$)",
```
Having the `|` here just complicates the test without providing any useful coverage of the non-test-code

cblecker replied:
Is it necessary for these regex matches to be so restrictive that even white space at end of line will cause the trigger to not fire? It's easy for white space to end up on a line (especially on mobile), and current functionality also allows for putting a comment like "k8s-bot something test this, flake ref:#999999"

fejta replied:
Actually I'm pretty sure the submit queue puts a bunch of stuff at the end, so we should explicitly validate that the current submit queue phrase continues to trigger.

crimsonfaith91 replied:
i will remove the unnecessary `|`, and add more test cases to handle corner cases. The suffix `\r?` does handle white spaces at end of line. i will double check the cases such as `/test\r\n flake ref:#999999`.

crimsonfaith91 commented:
As current implementation only takes care of white space(s) after the commands, i decided to rewrite all the regexes to accommodate the cases for multi-line and extra comments behind.

Here's the new format:
```
(1) The commands can have additional commands behind them, but not before
(2) Any additional comment and white spaces before the commands will NOT trigger the commands
(3) The commands can have comma appended, e.g., "ok to test," "/ok-to-test," "/test,"
```
Feel free to give feedback about this @fejta @cblecker

crimsonfaith91 commented:
I had a hard time working on the `/test something` regex that satisfies two regexps that contradict with each other below:
(1) having characters and/or white spaces after the regex
(2) having `/test` to trigger execution of _all_ always-run tests

For example, there are 3 tests: gce, unit, and gke. Writing `/test` should run all the 3 tests, but i am not able to find a regex that works for something like `/test gce`. Writing `/test gce` will still run all the 3 tests as there is no any `$` sign after the regex to differentiate the two cases.

As we want comma and white spaces after the regex _on the same line_, perhaps we can have something like `/test-gce`.
@fejta @cblecker

fejta commented:
How about `/test all\s+`

crimsonfaith91 commented:
@fejta great idea! let me reiterate to ensure i am on the same page with you. :)

For example, there are 3 tests: gce, unit, and gke.
(1) Writing `/test all` runs all the 3 tests. Writing `/test gce` runs only gce test.
(2) The regex handles multiple lines. The comment below runs all the tests.
```
Hello.
/test all
Thanks!
```
(3) _On the same line_, any comment or white space before commands will NOT trigger the commands. The comment below does NOT run any test.
```
Hello.
please /test all
Thanks!
```
(3) The commands can have comma appended. The comment below runs all the tests.
```
Hello.
/test all,
Thanks!
```

In other words, `/test` does not trigger any test anymore. Am i right?

crimsonfaith91 commented:
To summarize, the `test all` commands below trigger tests:
`/test all`
`/test all please`
`/test all,` (no whitespace character tailing)
`/test all,` (whitespace character tailing)
`/test all, please` (there is a space between `/test all,` and `please`)

The commands below do not trigger tests:
`please /test all`
`/test all,please` (there is no space between `/test all,` and `please`)

fejta commented:
Yep sounds great!

`/test` triggers nothing
`/test all` triggers everything and can have `/test all extra junk afterwards, assuming there's a space`
`/test foo` triggers the foo job

crimsonfaith91 commented:
Ya! In addition, "/test all," also works (a comma behind). Does it look
good to do so? :)

On Jun 14, 2017 11:08 PM, "Erick Fejta" <notifications@github.com> wrote:

> Yep sounds great!
>
> /test triggers nothing
> /test all triggers everything and can have /test all extra junk
> afterwards, assuming there's a space
> /test foo triggers the foo job
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/2891#issuecomment-308638189>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANRtXqPQ2fSVM-oosVpljXKMXt5tyVBSks5sEMpjgaJpZM4NrEox>
> .
>

fejta commented:
yep, perfect

crimsonfaith91 commented:
@fejta I have updated the regex to fit with the new requirement. I do not change `config.yaml` file yet - planning to wait for further review on current changes before modifying the file.

There are two questions regarding test cases:
(1) Should test cases in `pr_test.go` handle both old and new regex formats?
(2) The test cases for `ic_test.go` and `pr_test.go` seem redundant to `jobs_test.go`. Should we remove **some** of them?

fejta commented on mungegithub/mungers/submit-queue.go:
```diff
@@ -72,7 +72,7 @@ const (
 	_ = fmt.Print
 	// This MUST cause a RETEST of everything in the sq.RequiredRetestContexts
-	retestBody = fmt.Sprintf("@%s test this [submit-queue is verifying that this PR is safe to merge]", jenkinsBotName)
+	retestBody = fmt.Sprintf("/test [submit-queue is verifying that this PR is safe to merge]")
```
`/test all`

crimsonfaith91 replied:
ya! i miss it...

fejta commented on mungegithub/mungers/submit-queue_test.go:
```diff
@@ -1071,7 +1071,7 @@ func TestSubmitQueue(t *testing.T) {
 				json.NewDecoder(r.Body).Decode(c)
 				msg := *c.Body
-				if strings.HasPrefix(msg, "@"+jenkinsBotName+" test this") {
+				if strings.HasPrefix(msg, "/test") {
```
`/test all`

crimsonfaith91 replied:
sure!

fejta commented on prow/plugins/trigger/ic.go:
```diff
@@ -26,8 +26,14 @@ import (
+// The commands can have additional commands behind them, but not before
+// Any additional comment and white spaces before the commands will NOT trigger the commands
+// The commands can have comma appended, e.g., "ok to test," "/ok-to-test," "/test," "/test unit,"
+var oldRegex = `(?m)^(@k8s-bot )?ok to test,?(\s+|$)`
```
Let's not modify the oldRegex

crimsonfaith91 replied:
ok!

fejta commented on prow/config/config_test.go:
```diff
@@ -82,9 +82,10 @@ func TestContextMatches(t *testing.T) {
-			t.Errorf("%s in %s rerun_command: %s != expected: %s", repo, p.Name, p.RerunCommand, expected)
+		newCmd := fmt.Sprintf("/test %s", p.Name)
+		oldCmd := fmt.Sprintf("@k8s-bot %s test this", p.Name)
+		if p.RerunCommand != newCmd && p.RerunCommand != oldCmd {
```
We should require `RerunCommand` to match `newCmd`

crimsonfaith91 replied:
You are right - the `RerunCommand` does not need to handle `@k8s-bot test this`. This is because we only use `/test` for re-running the tests.

fejta replied:
The reason is because RerunCommand is the concrete example string we use to tell the user how to rerun the job.

The trigger regex is the only thing that triggers the job.

crimsonfaith91 replied:
noted! i recently use the RerunCommand a lot. :)

fejta commented on prow/plugins/trigger/ic.go:
```diff
@@ -26,8 +26,14 @@ import (
+// Any additional comment and white spaces before the commands will NOT trigger the commands
+// The commands can have comma appended, e.g., "ok to test," "/ok-to-test," "/test," "/test unit,"
+var oldRegex = `(?m)^(@k8s-bot )?ok to test,?(\s+|$)`
+var newRegex = `(?m)^/ok-to-test,?(\s+|$)`
```
Let's not support trailing whitespace in scenarios aside from `/test all`

crimsonfaith91 replied:
noted. i will remove the support for `ok-to-test` `/joke`, and `retest`.

fejta commented on prow/config/jobs_test.go:
```diff
@@ -76,8 +79,8 @@ func TestPresubmits(t *testing.T) {
 			// will attempt to rerun forever.
-			if job.AlwaysRun && !job.re.MatchString(testThis) {
-				t.Errorf("AlwaysRun job %s: \"%s\" does not match regex \"%v\".", job.Name, testThis, job.Trigger)
+			if job.AlwaysRun && !job.re.MatchString(newTestThis) && !job.re.MatchString(oldTestThis) {
```
Every job with `AlwaysRun` must match **all** of the following:
* `/test all EXTRA STUFF`
* `@k8s-bot test this EXTRA STUFF`
* `/test all`
* `@k8s-bot test this`

not just one of them

Also, if `job.AlwaysRun == false` we should validate it does not match any of them (aka `job.AlwaysRun == job.re.MatchString(newTestThisAndTheOtherOnes)`).

crimsonfaith91 replied:
yup! i will add the other conditions.

crimsonfaith91 replied:
To clarify, there will be two conditions to check:
```
if job.AlwaysRun && !job.re.MatchString(newTestThis) && !job.re.MatchString(oldTestThis)
if !job.AlwaysRun && (job.re.MatchString(newTestThis) || job.re.MatchString(oldTestThis))
```

First check ensures `AlwaysRun` jobs match _at least_ one of the commands.
Second check ensures `Non-AlwaysRun` jobs does not match _any_ of the commands.

fejta replied:
The boolean logic here is such that this test case will pass if either newTestThis or oldTestThis matches when AlwaysRun (try it out).

A cleaner way to do this might be to check:
```
`job.re.MatchString(newTestThis) == job.re.MatchString(oldTestThis)` // Do the new and old test do the same thing?
`job.AlwaysRun == job.re.MatchString(newTestThis)`  // Do we do the right thing based on AlwaysRun?
```
Although feel free to just fix up this line `if Always && !matchNew && !matchOld { error }` produces the wrong result when `matchNew=true` but `matchOld=false`

fejta commented on prow/config/jobs_test.go:
```diff
@@ -129,37 +132,122 @@ func TestCommentBodyMatches(t *testing.T) {
 		{
 			"org/repo",
-			"this is a random comment",
+			"/this-is-a-random-comment gce",
```
There is no need to change `TestCommentBodyMatches`, as this is purely a synthetic test with no real data. 

AKA if we don't update the `c := &Config{}` definition we also don't need to update the `var testcases = []struct{}` definition, saving us a bunch of churn

crimsonfaith91 replied:
aside from `/ok-to-test`, there are some tests for `/test all` too. for now, i will remove irrelevant tests for `/ok-to-test`.

crimsonfaith91 replied:
as discussed, i will revert _all_ changes to reduce churn.

fejta commented on prow/config/jobs_test.go:
```diff
@@ -168,38 +256,39 @@ func TestCommentBodyMatches(t *testing.T) {
 	}
 	for _, tc := range testcases {
-		actualJobs := c.MatchingPresubmits(tc.repo, tc.body, regexp.MustCompile(`ok to test`))
+		actualJobs := c.MatchingPresubmits(tc.repo, tc.body, regexp.MustCompile(`(?m)^/ok-to-test,?(\s+|$)`))
```
The only reason we need to update `testcases = []struct{}` above to have a `/ok-to-test` case instead of `ok to test` case is because we changed the definition of this regex here.

Same for changing `@k8s-bot federation test this` to `(?m)^/test federation,?(\s+|$)`

crimsonfaith91 replied:
do you mean we should still test for `ok to test` and `@k8s-bot test this` for `jobs_test.go`? i use test cases in this file to ensure changes to `config.yaml` works fine.

fejta replied:
But this doesn't have anything to do with validating `config.yaml`, since you're passing in the ok to test regex definition into c.MatchingPresubmits() here. It just validates the behavior of the MatchingPresubmits with fake data.

crimsonfaith91 replied:
ya - the presubmit and `config.yaml` are totally different things. Presubmit has the fake data - having the test cases passed at the presubmit using new syntax helps me to improve regex. :)

fejta commented on prow/plugins/trigger/ic.go:
```diff
@@ -26,8 +26,14 @@ import (
+var newRegex = `(?m)^/ok-to-test,?(\s+|$)`
+var regex = fmt.Sprintf("(%s|%s)", oldRegex, newRegex)
+var okToTest = regexp.MustCompile(regex)
+var retest = regexp.MustCompile(`(?m)^/retest,?(\s+|$)`)
```
I don't see any need to change the retest command in this PR

crimsonfaith91 replied:
noted! i will remove the support and related test cases.

fejta commented on prow/plugins/trigger/ic.go:
```diff
@@ -26,8 +26,14 @@ import (
+var oldRegex = `(?m)^(@k8s-bot )?ok to test,?(\s+|$)`
+var newRegex = `(?m)^/ok-to-test,?(\s+|$)`
+var regex = fmt.Sprintf("(%s|%s)", oldRegex, newRegex)
+var okToTest = regexp.MustCompile(regex)
```
Do we really need four variables for this? seems like we could just do `regexp.MustComplile('ok to test|(?m)^/ok-to-test\r?$')`

crimsonfaith91 replied:
as we will not support trailing whitespaces, comma, and words for `ok-to-test`, a line suffices.

crimsonfaith91 commented on prow/plugins/trigger/ic_test.go:
```diff
@@ -107,15 +131,79 @@ func TestHandleIssueComment(t *testing.T) {
 		{
 			Author:      "t",
-			Body:        "@k8s-bot test this",
+			Body:        "/test all",
```
i didn't remove these test cases for `/test all` here because the duplicated test cases have been removed from `jobs_test.go`.

fejta replied:
These additional test cases do not add any test coverage. They just test the definition of the fake regex data you define below.

Please choose a simpler regex to make a simpler test with the same coverage

crimsonfaith91 commented on prow/plugins/trigger/ic.go:
```diff
@@ -26,7 +26,10 @@ import (
+// The commands can have additional commands behind them, but not before
+// Any additional comment and white spaces before the commands will NOT trigger the commands
+// The commands can have comma appended, e.g., "ok to test," "/ok-to-test," "/test," "/test unit,"
+var okToTest = regexp.MustCompile(`((?m)^(@k8s-bot )?ok to test\r?$|(?m)^/ok-to-test\r?$)`)
```
The existing codes have the regex as `(?m)^(@k8s-bot )?ok to test\r?$`. i left it intact.

fejta commented on prow/config/jobs_test.go:
```diff
@@ -76,8 +79,12 @@ func TestPresubmits(t *testing.T) {
+				t.Errorf("AlwaysRun job %s: \"%s\" and \"%s\" do not match regex \"%v\".", job.Name, newTestThis, oldTestThis, job.Trigger)
+			}
+			// Check that the merge bot will not run Non-AlwaysRun jobs
+			if !job.AlwaysRun && (job.re.MatchString(newTestThis) || job.re.MatchString(oldTestThis)) {
```
FYI this part works.

crimsonfaith91 replied:
sorry for late reply. thanks for the clear explanation!

I will change
```
`if job.AlwaysRun && !job.re.MatchString(newTestThis) && !job.re.MatchString(oldTestThis)`
```
to
```
`job.re.MatchString(newTestThis) == job.re.MatchString(oldTestThis)`
`job.AlwaysRun == job.re.MatchString(newTestThis)`
```

crimsonfaith91 replied:
Sorry, i am still confused why `if job.AlwaysRun && !job.re.MatchString(newTestThis) && !job.re.MatchString(oldTestThis)` does not work.
This check will only trigger the error message if _both_ new and old regexes do not get matched.
In other words, if either one of the regex gets matched, the check will pass.
This is coherent with the requirement for having the check passes if either newTestThis or oldTestThis matches when AlwaysRun.

fejta replied:
But we want to validate that **both** newTestThis and oldTestThis matches when AlwaysRun.

crimsonfaith91 replied:
i think i got the idea finally! 
The check should trigger the error message if **either** of the regexes do not get matched.
I will have the condition as: `if job.AlwaysRun && (!job.re.MatchString(newTestThis) || !job.re.MatchString(oldTestThis))`

fejta commented on prow/plugins/trigger/ic_test.go:
```diff
@@ -174,14 +288,14 @@ func TestHandleIssueComment(t *testing.T) {
 					AlwaysRun: true,
 					Context:   "pull-job",
-					Trigger:   "@k8s-bot test this",
+					Trigger:   `(?m)^/test all,?(\s+|$)`,
```
This is the regex that defines what fake jobs exist and what fake regexes trigger these fake jobs. This is totally fake data that has no relevance on our actual config.

crimsonfaith91 replied:
yup! i think this is the best explanation regarding choosing betwee test coverage and churn reduction. to simplify, i will only test regex `^/test all$` (without comma, multiple line, and extra spaces behind).

fejta commented on prow/config.yaml:
```diff
@@ -18,36 +18,36 @@ presubmits:
-    rerun_command: "@k8s-bot pull-cadvisor-e2e test this"
-    trigger: "@k8s-bot (pull-cadvisor-e2e )?test this"
+    rerun_command: "/test pull-cadvisor-e2e"
+    trigger: "(^@k8s-bot (pull-cadvisor-e2e )?test this\r?$|^/test( pull-cadvisor-e2e)?\r?$)"
```
I don't think we want the `\r$` suffix (see next comment)

crimsonfaith91 replied:
i haven't updated the regex for `config.yaml` yet. i am waiting for reply regarding `if job.AlwaysRun && !job.re.MatchString(newTestThis) && !job.re.MatchString(oldTestThis)`.

crimsonfaith91 replied:
The new regex is: 
`((?m)^@k8s-bot (pull-cadvisor-e2e )?test this,?(\\s+|$)|(?m)^/test( all| pull-cadvisor-e2e),?(\\s+|$))`

It is long, but IMO once we retired the old commands, it is readable.

fejta commented on prow/config/jobs_test.go:
```diff
@@ -28,7 +28,10 @@ import (
 
-const testThis = "@k8s-bot test this"
+const (
+	newTestThis = "/test all"
```
Might want to also add these two:
```
retestBody = fmt.Sprintf("/test all [submit-queue is verifying that this PR is safe to merge]")
retestBody = fmt.Sprintf("@k8s-bot test this [submit-queue is verifying that this PR is safe to merge]")
```

I suspect you'll find that they fail

crimsonfaith91 replied:
noted! will add more conditions on them :)

crimsonfaith91 replied:
i will add both of them at `submit-queue.go`.

fejta replied:
I think adding a `"/test this [extra stuff]"` check is also worthwhile

fejta commented on mungegithub/mungers/submit-queue.go:
```diff
@@ -72,7 +72,7 @@ const (
 var (
 	_ = fmt.Print
 	// This MUST cause a RETEST of everything in the sq.RequiredRetestContexts
-	retestBody = fmt.Sprintf("@%s test this [submit-queue is verifying that this PR is safe to merge]", jenkinsBotName)
```
This is the comment that today's submit queue leaves to trigger testing before merging.

We need to ensure that all AlwaysRun jobs trigger on this comment.

crimsonfaith91 replied:
noted! i will need to have both new and old retest body. i will make the relevant changes.

fejta commented on mungegithub/mungers/submit-queue.go:
```diff
@@ -72,7 +72,7 @@ const (
 	_ = fmt.Print
 	// This MUST cause a RETEST of everything in the sq.RequiredRetestContexts
-	retestBody = fmt.Sprintf("@%s test this [submit-queue is verifying that this PR is safe to merge]", jenkinsBotName)
+	retestBody = fmt.Sprintf("/test all [submit-queue is verifying that this PR is safe to merge]")
```
This is the comment that tomorrow's submit queue will leave to trigger testing before merging.

We need to ensure that all AlwaysRun jobs trigger on this comment as well.

crimsonfaith91 replied:
yup! both should work after the merge.

crimsonfaith91 commented:
@fejta Finally updated `config.yaml`!

I am not sure whether we should have tests below at `jobs_test.go`:
```
newRetestBody = "/test all [submit-queue is verifying that this PR is safe to merge]"
oldRetestBody = fmt.Sprintf("@k8s-bot test this [submit-queue is verifying that this PR is safe to merge]")
```

The file before change does not have the test for `@k8s-bot test this [submit-queue is verifying that this PR is safe to merge]`

fejta commented on mungegithub/mungers/submit-queue.go:
```diff
@@ -72,7 +72,8 @@ const (
 	// This MUST cause a RETEST of everything in the sq.RequiredRetestContexts
-	retestBody = fmt.Sprintf("@%s test this [submit-queue is verifying that this PR is safe to merge]", jenkinsBotName)
+	newRetestBody = "/test all [submit-queue is verifying that this PR is safe to merge]"
+	oldRetestBody = fmt.Sprintf("@%s test this [submit-queue is verifying that this PR is safe to merge]", jenkinsBotName)
```
Good catch, although I don't consider leaving one extra comment on a PR an important problem to solve.

If you want to keep this please add a TODO to delete this?

crimsonfaith91 replied:
yes, having a TODO to remove `oldRetestBody` when the time comes is a good suggestion. :)

fejta commented on mungegithub/mungers/submit-queue.go:
```diff
@@ -577,7 +578,7 @@ func (sq *SubmitQueue) AddFlags(cmd *cobra.Command, config *github.Config) {
 		"Comma separated list of jobs in Jenkins to use for stability testing that needs only weak success")
 	cmd.Flags().StringSliceVar(&sq.RequiredStatusContexts, "required-contexts", []string{}, "Comma separate list of status contexts required for a PR to be considered ok to merge")
-	cmd.Flags().StringVar(&sq.RetestBody, "retest-body", retestBody, "message which, when posted to the PR, will cause ALL `required-retest-contexts` to be re-tested")
+	cmd.Flags().StringVar(&sq.RetestBody, "retest-body", newRetestBody, "message which, when posted to the PR, will cause ALL `required-retest-contexts` to be re-tested")
```
Does this flag even work? It looks like we always use newRetestBody regardless of the flag's value.

crimsonfaith91 replied:
this flag exists before the change (when there is only one type of `retestBody`). i don't think we should remove it, but i am not sure about this.

fejta replied:
It isn't searching for "retest-body" doesn't show anything. It is dead code.

Can you delete the sq.RetestBody from here and then ensure unit tests continue to pass:

https://github.com/kubernetes/test-infra/blob/e91799928ba3c4d7ee791dbd1316042067f12c2f/mungegithub/mungers/submit-queue.go#L224

crimsonfaith91 replied:
got it! i will remove `RetestBody` attribute of the submit queue.

fejta commented:
Please squash commits into one

crimsonfaith91 commented:
@fejta added the tests for `newRetestBody` and `oldRetestBody`. squashed the commits too.

fejta commented:
/lgtm

We should eliminate the flag that doesn't actually do anything.

crimsonfaith91 commented:
@fejta removed the flag and squashed. every feedback has been addressed. :)

crimsonfaith91 commented:
Sure! I also need to send an email to kubernetes-dev@ informing the change. :)

spxtr commented:
What was the reason to do `/test all` to trigger all tests instead of just `/test`? What if we add a job named `all` :smiling_imp:

crimsonfaith91 commented:
@spxtr IMO, `/test all` will make it clear that we are running all tests. Having `test` along will make people think that it is relevant to `retest`.

I also had a hard time working on `/test <some-test-name>` regex that satisfies two cases that contradict with each other below:
(1) having characters, comma, and/or white spaces after the regex
(2) having `/test` to trigger execution of all always-run tests

For example, there are 3 tests: gce, unit, and gke. Writing `/test` should run all the 3 tests, but i was not able to find a regex that worked for something like `/test gce`. Writing `/test gce` will still run all the 3 tests as `/test` is substring of `/test gce`. Having `test all` solve the problem and make the regex shorter.

Nevertheless, i believe there should be a way to come out with a concise, robust regex satisfying the two requirements.

spxtr commented:
> (1) having characters, comma, and/or white spaces after the regex

For most of our commands that won't work. If I say `/lgtm, words words words` then no `lgtm` label will show up. This was intended to make it easier to write plugins since you don't need to think about this sort of thing.

With that in mind, `/test` on its own line should run all tests. `/test pull-k8s-gce` won't match the test-all regex, since there's other junk before the end.

crimsonfaith91 commented:
@spxtr ya, but i was told that we need to consider situation like `/test all, words words words`. i spent a considerable amount of time working on handling this.
@cblecker do you mind to verify that we need to consider situation like `/test all, words words words`? i remembered you mentioned about this in a review comment, but i squashed the PR and could not find it.

cblecker commented:
@crimsonfaith91 My comment was here:
https://github.com/kubernetes/test-infra/pull/2891#discussion_r121572881
> Is it necessary for these regex matches to be so restrictive that even white space at end of line will cause the trigger to not fire? It's easy for white space to end up on a line (especially on mobile), and current functionality also allows for putting a comment like "k8s-bot something test this, flake ref:#999999

While @spxtr is right that this may not work for LGTM, it's current functionality with the k8s-bot commands. Additionally, the submit queue today issues the following command to get prow to kick off the second set of e2es:
```
@k8s-bot test this [submit-queue is verifying that this PR is safe to merge]
```

crimsonfaith91 commented:
@cblecker thanks for the clarification! yup, i have both old and new commands work for submit queue. :)
```
@k8s-bot test this [submit-queue is verifying that this PR is safe to merge] 
/test all [submit-queue is verifying that this PR is safe to merge]
```
